### PR TITLE
Add state.show_states call.

### PR DESCRIPTION
**Pull request by epcim**
### What does this PR do?
list individual sls state's to be applied.

### What issues does this PR fix or reference?
I found the function under this discussing: https://github.com/saltstack/salt/issues/6265
Thx, @cenkalti for writing it.

### Tests written?

No

### Commits signed with GPG?

Yes

### How this can be used
One can get the list of these low state's to be applied and then query how many of them would perform a change on a node.  For example, this is today script (https://github.com/salt-formulas/salt-formulas-scripts/blob/master/salt-state-apply-trend.sh) I am using.

(as up to now we don't have any statistics of this kind available (feel free to correct me)).

### CODE REVIEW

**Comment by garethgreenaway** [comment]
@epcim Great addition!  Thanks!  Would you mind adding some unit tests to the PR for this new function?  Thanks!

File: salt/modules/state.py [hunk 1]
```diff
@@ -1352,6 +1353,20 @@ def show_state_usage(queue=False, **kwargs):
     return ret
 
 
+def show_states():
+    """Returns the list of states that will be applied on highstate."""
```
Comment (cachedout) [inline salt/modules/state.py]: Could you please break these out into three lines and change the double-quotes to single-quotes? Thanks.

**Review by cachedout — CHANGES_REQUESTED** [review]

**Comment by cachedout** [comment]
@garethgreenaway @epcim Integration tests might be a better choice here, IMHO.

File: salt/modules/state.py (hunk 1 above)
Comment (rallytime) [reply salt/modules/state.py]: This could also use a `.. versionadded:: Oxygen` tag while you're editing this.

**Comment by epcim** [comment]
I am not satisfied with the speed. Anyone know, how we could modify `compile_high_data` to speed up the results, as we are only interested in sorted list of `__sls__` values...

**Review by garethgreenaway — COMMENTED** [review]
Looks good to me but @cachedout had requested integration tests in addition to or instead of the unit tests, so I'll defer.

**Comment by rallytime** [comment]
Ah, yes. Integration tests would be good here @epcim. Please let us know if you need any help.

**Comment by epcim** [comment]
[**@rallytime**](https://github.com/rallytime) commented on [Dec 5, 2017, 10:39 PM GMT+1](https://github.com/saltstack/salt/pull/44475#issuecomment-349450140 "2017-12-05T21:39:03Z - Replied by Github Reply Comments"):
> Ah, yes. Integration tests would be good here [@epcim](https://github.com/epcim). Please let us know if you need any help.

I dont have any idea how to write tests properly for that, as what it does it's 1:1 code from low_sls call (if I remember + additional 5 lines of loop in order to pick __sls__. That code is supposed to be already tested by mentioned method.

The only test I can imagine is that it returns a type list. As the in/out vlaues will be in the test call anyway modified to match. If you really need tests, can you post an update with some example that I could work on. 

Worth if you point out why it;s failing..

**Comment by rallytime** [comment]
@epcim Thank you so much for adding these tests. I just checked and they're passing nicely. The other failures here are not related.

@garethgreenaway Can you swing by here one more time when you get a moment?

**Comment by epcim** [comment]
Squashed commit's in the last push.
+ removed the unsorted workflow with modified compile_high* as not faster.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Review by garethgreenaway — APPROVED** [review]

**Comment by epcim** [comment]
Failures seems to me are not related to this patch..

**Comment by max-arnold** [comment]
@epcim How this is different from `state.show_top`?

**Comment by cenkalti** [comment]
@max-arnold `show_top` lists states only defined in `top.sls` whereas `show_states` lists all states that will be applied to minion.

In other words, if a state is included from another state, it will not be in `show_top` output but be in `show_states` output.

**Pull request merged**